### ISSUE TITLE
Improve build format for Debian/Ubuntu

### ISSUE THREAD

**Issue by stoecker** [openSUSE/open-build-service#9744]
When building for RPM based systems only spec file and changes file must be modified. When building for DEB based formats multiple files must be modified in a binary blob. That's awful.

For I our internal builds I thus created a "*.debian" fileformat to use within OBS build.

It looks like this:

```
%define debug           0
%define makedocs        1
%define notls           1
%if %{repository} eq "Debian_5.0"
%define nousb           1
%define libusb
%else
%define nousb           0
%define libusb          libusb-1.0-0-dev,
%endif
%if %{repository} eq "Debian_5.0" || %{repository} eq "Ubuntu_14.04"
%define makedocs        0
%define tex
%else
%define tex             texlive-lang-german, texlive-latex-base, \
  texlive-fonts-recommended, texlive-fonts-extra, texlive-latex-extra, \
  texlive-bibtex-extra, texlive-latex-recommended, texlive-generic-extra,
%endif
%define requires perl, %{tex} %{libusb} ghostscript, netpbm, libssl-dev
%define _bindir         /usr/bin
....
%define buildroot       /usr/src/packages/BUILD/debian/%{name}/
%%%%%%{name}.dsc%%%%%
Format: 1.0
Source: %{name}
Version: %{version}
Binary: %{name}
Maintainer: %{packager}
Architecture: any
Build-Depends: debhelper (>= %{debhelperversion}), %{requires}
Files: 
 %{releasemd5} releasesource.tar.gz
%%%%%debian.tar.gz/debian/changelog%%%%%
%{name} (2.0) stable; urgency=low

  * First Debian version

 -- Name <mail@mail.mail>  Thu, 24 Nov 2016 00:00:08 +0100
%%%%%debian.tar.gz/debian/control%%%%%
# https://www.debian.org/doc/debian-policy/ch-controlfields.html
Source: %{name}
Section: Productivity/Other
Priority: optional
Maintainer: %{packager}
Build-Depends: debhelper (>= %{debhelperversion}), %{requires}

Package: %{name}
Architecture: any
Depends: ${shlibs:Depends},
  perl,
  cron
Description: description text
%%%%%debian.tar.gz/debian/compat%%%%%
%{debhelperversion}
%%%%%debian.tar.gz/debian/rules%%%%%
#!/usr/bin/make -f

# Uncomment this to turn on verbose mode.
#export DH_VERBOSE=1

# This is the debhelper compatibility version to use.
export DH_COMPAT=%{debhelperversion}

LIBDIR = /usr/lib/

CFLAGS = -g
ifneq (,$(findstring noopt,$(DEB_BUILD_OPTIONS)))
CFLAGS += -O0
else
CFLAGS += -O2
endif

OPT =
ARG =
%if %{notls} == 1
ARG+= NOTLS=1
%endif
...
%%%%%debian.tar.gz/debian/%{name}.preinst%%%%%
#!/bin/sh
/usr/sbin/useradd -m -r -d %{homdir} -s /bin/bash -c "%{username}" -g %{ugroup} %{user} 2> /dev/null || :
%%%%%debian.tar.gz/debian/%{name}.postinst%%%%%
...
%%%%%debian.tar.gz/debian/%{name}.postrm%%%%%
...
```

Functions are:
* %%%%%..%%%%% creates a file inside debian.tar.gz
* %define creates a variable
* variables repository, repositoryname, repositoryversion, debhelperversion, arch - are passed by call-script
* %if, %elif, %else, %endif are for checks like in RPM spec file, they must be resolvable with the provided build environment to select correct parts
* There should be an extension %{changelog} to import and convert a *.changes file.

Now I have a simple parser, which sets the variables, handles the ifs and if this preprocessor step is done the splits the file into the individual parts and creates a debian.tar.gz, which then is used for building.

This way I have only one file containing all the build information like for spec files and this is non-binary, so it can be tracked with changes.

In can provide the parser, but that's about 150 lines of straightforward perl code, nothing special.

I think it would be a big benefit, when OBS would support this or a similar fileformat directly.

I'm using this for about 3 years now and it works like a charm.

**Comment by adrianschroeter** [comment]
you can already have files like debian.control which ends up as debian/control in the build system. So you do not have to use the tar balls. is that solving your issue?

**Comment by stoecker** [comment]
No. That does not solve the conditionals, the single file approach and variable definition and usage. The only advantage of having the single files compared to the debian file is the possibility of change tracking (with the large disadvantage of having many files).

If you try to build for different Debian/Ubuntu version from one source you very fast step into problems.

**Comment by nickbroon** [comment]
The fact that debtransform is also limited to 1.0 style Debian packages (as compared to now more widely used 3.0) causes issues. Debtransform does not support variable substitution.

https://github.com/openSUSE/obs-build/issues/500
https://github.com/openSUSE/obs-build/issues/369

**Comment by stoecker** [comment]
If I understand that correctly my approach would also fix openSUSE/obs-build#500 and openSUSE/obs-build#369 as debtransform is not needed.

I'd be willing to help with the implementation if this approach gets agreed on.

From the report openSUSE/obs-build#500 I take the fact that there should be a way to include binary files into the created debian.tar.gz archive, i.e. by specifying something like the Source lines of RPM spec format.

**Comment by Conan-Kudo** [comment]
@stoecker You can also just use spec files to build Debian packages... That's been supported for many years now.

**Comment by Conan-Kudo** [comment]
Here's an example: https://build.opensuse.org/package/show/home:Pharaoh_Atem:Debian_Pagure/pagure
